### ISSUE TITLE
Performance of mutating large collections: arrays vs objects

### ISSUE THREAD

**Issue by keriwarr** [immerjs/immer#649]
## 🙋‍♂ Question

Hello!

When I first read the immer [performance docs](https://immerjs.github.io/immer/docs/performance), it didn't occur to me at the time that choice of Array vs Object for storing data might make a large impact on performance.

I use `immer` to manage a redux store in which [records are keyed by their ID inside of an object](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape#designing-a-normalized-state). This performs perfectly well when the number of records is in the low 1000s, however we are looking to push our application to the level of supporting tens or possibly even hundreds of thousands of records. What I've found is that for sufficiently large collections it can take _vastly_ longer to update **a single record** if the collection is an object as opposed to an array.

1) Is there any known workaround such that I can get array-like performance, but continue to use an object? (Or some other data structure that gives us O(1) retrieval by ID?)
2) If there is no such workaround, would you accept a PR to the docs clarifying this?

## Link to repro

https://codesandbox.io/s/immer-sandbox-g64y4

In this demo, I create an array collection and an object collection of todos (just like in the `test:perf` code) but each with size 200,000, and then measure the time it takes to update a single record within the collection.

![image](https://user-images.githubusercontent.com/5460144/88980511-e5759280-d291-11ea-905a-add8265e565b.png)

**Update**: `Map`s seem to be _about_ an order of magnitude faster than objects but still a few times slower than arrays (I tried out a few different collection sizes). However, I don't think that I would want to take the hit of using an unserializable data structure in redux.

https://codesandbox.io/s/immer-sandbox-09zyb

![image](https://user-images.githubusercontent.com/5460144/88982830-4ce21100-d297-11ea-8649-406fe310cb6f.png)

## Environment

We only accept questions against the latest Immer version.

- **Immer version:** 7.0.7
- [x] Occurs with `setUseProxies(true)`
- [x] Occurs with `setUseProxies(false)` (ES5 only)

P.S. thanks for this awesome package <3

**Comment by mweststrate** [comment]
Sorry, didn't really have time to look into it so far. Some things to check:

1. autofreeze is no longer faster by default, and actually slower in many cases (the docs are outdated in this regard). Actually keeping the state frozen is often faster on the long term as immer can bailout faster on frozen objects when looking for changes in the draft
2. if your data structure is so large that the immer overhead starts to count, you are always fighting an uphill battle, even if things can be improved (feel free to investigate where the surprising difference is coming from). I'd skip drafting for such a perf critical large object, and rather only draft the individual item you are about to update, and then merge it back into the large collection using the vanilla js. (so you could wrap only sub reducers in produce, without using it for the root reducer for example)

**Comment by keriwarr** [comment]
thanks Michel, I'll drop a comment here when I arrive at a solution I'm happy with.

**Comment by bfelbo** [comment]
This is the top result on Google for Immer arrays vs. objects benchmarking so sharing info here in case it's useful for future readers.

Thanks @keriwarr for the benchmarking scripts. I just tried them with Immer 10 to see if there's still a performance difference between array and object. Seems like there is, but much reduced.

I just changed Immer to 10.0.3 and removed `setUseProxies(true)`, which was failing. Your benchmark now produces:

```
[ARRAY] immer (proxy) - without autofreeze: 9ms 
[OBJECT] immer (proxy) - without autofreeze: 34ms 
[MAP] immer (proxy) - without autofreeze: 14ms 
```

and

```
[OBJECT] immer (proxy) - without autofreeze: 7ms 
[OBJECT] immer (proxy) - with autofreeze: 15ms
[ARRAY] immer (proxy) - without autofreeze: 2ms 
[ARRAY] immer (proxy) - with autofreeze: 24ms
```

so it seems like it's still worth using arrays for large performance-critical applications that need Immer. Hope this is helpful for anyone else looking into Immer performance.
